ee/localserver/krypto-ec-middleware_test.go

```diff
@@ -141,116 +75,190 @@ func TestKryptoEcMiddleware(t *testing.T) {
+			if tt.expectedPresenceDetectionCallbackHeaders != nil {
+				callbackWaitGroup.Add(2)
 			}
 
-			responseMap[testMsgKey] = responseValue
+			// this is the key we can use to open the response
+			// it gets set later
+			var challengePrivateKey *[32]byte
 
-			responseDataRaw := mustMarshal(t, responseMap)
+			callbackServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				defer callbackWaitGroup.Done()
 
-			testHandler := tt.handler
+				body, err := io.ReadAll(r.Body)
+				require.NoError(t, err)
 
-			// this handler is what will respond to the request made by the kryptoEcMiddleware.Wrap handler
-			// in this test we just want it to regurgitate the response data we defined above
-			// this should match the responseData in the opened response
-			if testHandler == nil {
-				testHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				opened := mustOpenKryptoResponse(t, mustExtractJsonProperty[string](t, body, "Response"), challengePrivateKey)
 
-					// make sure all the request headers are present
-					for k, v := range cmdRequestHeaders {
-						require.Equal(t, v[0], r.Header.Get(k))
-					}
+				headers := mustExtractJsonProperty[map[string][]string](t, opened.ResponseData, "headers")
 
-					reqBodyRaw, err := io.ReadAll(r.Body)
-					require.NoError(t, err)
-					defer r.Body.Close()
+				// assume that if we have presence detection header, this is the presence detection callback
+				if _, ok := headers[kolideDurationSinceLastPresenceDetectionHeaderKey]; ok {
+					require.Equal(t, tt.expectedPresenceDetectionCallbackHeaders, headers,
+						"presence detection callback headers should match expected",
+					)
+					return
+				}
 
-					require.Equal(t, cmdReqBody, reqBodyRaw)
-					w.Write(responseDataRaw)
-				})
+				require.Equal(t, tt.expectedCallbackHeaders, headers,
+					"callback headers should match expected",
+				)
+			}))
+
+			cmdReq := v2CmdRequestType{
+				Path:            "whatevs",
+				Body:            []byte(randomStringWithSqlCharacters(t, 100000)),
+				Headers:         tt.cmdReqHeaders,
+				CallbackUrl:     callbackServer.URL,
+				CallbackHeaders: tt.cmdReqCallbackHeaders,
 			}
 
-			challengeBytes, privateEncryptionKey := tt.challenge()
+			challengeKryptoBoxB64, challengePrivateKey := mustGenerateChallenge(t, remoteServerPrivateKey, challengeId, challengeData, mustMarshal(t, cmdReq))
+
+			responseBody := mustMarshal(t, map[string]any{
+				"body": ulid.New(),
+			})
 
-			encodedChallenge := base64.StdEncoding.EncodeToString(challengeBytes)
-			for _, req := range []*http.Request{makeGetRequest(t, encodedChallenge), makePostRequest(t, encodedChallenge)} {
+			for _, req := range []*http.Request{mustMakeGetRequest(t, challengeKryptoBoxB64), mustMakePostRequest(t, challengeKryptoBoxB64)} {
 				req := req
 				t.Run(req.Method, func(t *testing.T) {
-					t.Parallel()
+					// t.Parallel()
```

Comment:
when these are made parallel it ends up waiting at the `callbackWaitGroup.Wait()` after this loop forever ... hmm